### PR TITLE
Use IRuntimeEnvironment for PlatformHelper

### PR DESCRIPTION
**Pull request by BrennanConroy**
@victorhurdugaci @davidfowl

### CODE REVIEW

File: src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs [hunk 1]
```diff
@@ -2,18 +2,23 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using Microsoft.Framework.Runtime.Infrastructure;
 
 namespace Microsoft.Framework.Runtime
 {
     internal static class PlatformHelper
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs]: Noooo, get rid of this and stop using CallContextServiceLocator everywhere :smile:

Comment (BrennanConroy) [inline src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs]: What am I supposed to use instead D:<

File: src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs [hunk 2]
```diff
@@ -2,18 +2,23 @@
...
+                    .Locator
+                    .ServiceProvider
+                    .GetService(typeof(IRuntimeEnvironment))).OperatingSystem);
 
         public static bool IsMono
         {
             get
             {
-                return _isMono.Value;
+                return _osName.Value == "Darwin";
```
Comment (moozzyk) [inline src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs]: forgive my ignorance but how this is supposed to work when we have CoreClr running on non-windows OSes?

Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs]: This check is wrong :smile:

File: src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs]: Use `IRuntimeEnvironment` in place of PlatformHelper

File: src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs (hunk 2 above)
Comment (BrennanConroy) [inline src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs]: Oh wow, I wasn't thinking D:

File: src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs (hunk 1 above)
Comment (BrennanConroy) [inline src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs]: IServiceProvider unfortunately isn't easily accessible to a lot of the locations that use PlatformHelper. For the places it is, something similar to ```if((((IRuntimeEnvironment)_services.GetService(typeof(IRuntimeEnvironment))).RuntimeType == "Mono"))``` will need to be used. And the places it isn't available we will need to use CallContextServiceLocator.

**Comment by BrennanConroy** [comment]
@davidfowl @ChengTian 
Passing services everywhere

File: src/Microsoft.Framework.Runtime.Hosting/Dependencies/FrameworkReferenceResolver.cs [hunk 3]
```diff
@@ -162,10 +164,10 @@ public static string GetReferenceAssembliesPath()
                     "Reference Assemblies", "Microsoft", "Framework");
         }
 
-        
         private FrameworkInformation GetFrameworkInformation(NuGetFramework targetFramework)
         {
-            string referenceAssembliesPath = GetReferenceAssembliesPath();
+            var isMono = ((IRuntimeEnvironment)_services.GetService(typeof(IRuntimeEnvironment))).RuntimeType == "Mono";
```
Comment (ChengTian) [inline src/Microsoft.Framework.Runtime.Hosting/Dependencies/FrameworkReferenceResolver.cs]: I see this several times. Can we encapsulate it with something like
```c#
SomeUtils.IsMono(IServiceProvider)
```

Comment (ChengTian) [inline src/Microsoft.Framework.Runtime.Hosting/Dependencies/FrameworkReferenceResolver.cs]: And `IsWindows` as well

File: src/Microsoft.Framework.PackageManager/Scripts/ScriptExecutor.cs [hunk 4]
```diff
@@ -15,7 +15,7 @@ public class ScriptExecutor
     {
         private static readonly string ErrorMessageTemplate = "The '{0}' script failed with status code {1}.";
 
-        public bool Execute(Runtime.Project project, string scriptName, Func<string, string> getVariable)
+        public bool Execute(Runtime.Project project, string scriptName, Func<string, string> getVariable, bool isMono)
         {
```
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Scripts/ScriptExecutor.cs]: Create a constructor taking `isMono` as a parameter.

File: src/Microsoft.Framework.PackageManager/Publish/PublishRuntime.cs [hunk 5]
```diff
@@ -11,11 +11,13 @@ public class PublishRuntime
     {
         private readonly FrameworkName _frameworkName;
         private readonly string _runtimePath;
+        private readonly bool _isMono;
 
-        public PublishRuntime(PublishRoot root, FrameworkName frameworkName, string runtimePath)
+        public PublishRuntime(PublishRoot root, FrameworkName frameworkName, string runtimePath, bool isMono)
```
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Publish/PublishRuntime.cs]: I guess it's better to have `PublishRoot.IsMono` instead of passing it in this way.

**Comment by ChengTian** [comment]
Need rebase

File: test/Microsoft.Framework.ApplicationHost.FunctionalTests/AppHostTests.cs [hunk 6]
```diff
@@ -135,7 +135,7 @@ public void AppHostShowsErrorWhenCurrentTargetFrameworkWasNotFoundInProjectJson(
             var runtimeTargetFramework = new FrameworkName(runtimeTargetFrameworkString);
             var runtimeTargetFrameworkShortName = VersionUtility.GetShortFrameworkName(runtimeTargetFramework);
             var runtimeType = flavor == "coreclr" ? "CoreCLR" : "CLR";
-            runtimeType = PlatformHelper.IsMono ? "Mono" : runtimeType;
+            runtimeType = TestUtils.CurrentRuntimeEnvironment.RuntimeType == "Mono" ? "Mono" : runtimeType;
```
Comment (BrennanConroy) [inline test/Microsoft.Framework.ApplicationHost.FunctionalTests/AppHostTests.cs]: Missed one

File: test/Microsoft.Framework.CommonTestUtils/TestUtils.cs [hunk 7]
```diff
@@ -157,8 +157,8 @@ public static DisposableDir GetRuntimeHomeDir(string flavor, string os, string a
             runtimeHomePath = CreateTempDir();
             runtimeName = Path.GetFileNameWithoutExtension(runtimeNupkg);
             var runtimeRoot = Path.Combine(runtimeHomePath, "runtimes", runtimeName);
-
-            if (!PlatformHelper.IsMono)
+
```
Comment (BrennanConroy) [inline test/Microsoft.Framework.CommonTestUtils/TestUtils.cs]: Remove space

File: test/Microsoft.Framework.PackageManager.FunctionalTests/DnuWrapTests.cs [hunk 8]
```diff
@@ -276,7 +280,7 @@ public void DnuWrapCreatesGlobalJsonIfSolutionRootDoesNotHaveOne(string flavor,
         {
             var runtimeHomeDir = TestUtils.GetRuntimeHomeDir(flavor, os, architecture);
 
-            if (PlatformHelper.IsMono)
+            if (TestUtils.CurrentRuntimeEnvironment.RuntimeType == "Mono")
```
Comment (BrennanConroy) [inline test/Microsoft.Framework.PackageManager.FunctionalTests/DnuWrapTests.cs]: Missed another

File: test/Microsoft.Framework.PackageManager.FunctionalTests/Microsoft.Framework.PackageManager.FunctionalTests.xproj [hunk 9]
```diff
@@ -13,5 +13,8 @@
   <PropertyGroup>
     <SchemaVersion>2.0</SchemaVersion>
   </PropertyGroup>
+  <ItemGroup>
```
Comment (BrennanConroy) [inline test/Microsoft.Framework.PackageManager.FunctionalTests/Microsoft.Framework.PackageManager.FunctionalTests.xproj]: Not sure, will remove

File: test/Microsoft.Framework.Runtime.Roslyn.Tests/ProjectExtensionsFacts.cs [hunk 10]
```diff
@@ -176,7 +177,7 @@ public static CompilationSettings GetCompilationSettings(this Project project, s
         {
             var framework = FrameworkNameHelper.ParseFrameworkName(frameworkName);
             return project.GetCompilerOptions(framework, "Debug")
-                          .ToCompilationSettings(framework);
+                          .ToCompilationSettings(framework, CallContextServiceLocator.Locator.ServiceProvider);
```
Comment (BrennanConroy) [inline test/Microsoft.Framework.Runtime.Roslyn.Tests/ProjectExtensionsFacts.cs]: Use common test utils

File: src/Microsoft.Framework.PackageManager/Install/AppCommandsFolderRepository.cs [hunk 11]
```diff
@@ -66,7 +68,7 @@ public NuGet.PackageInfo FindCommandOwner(string command)
         public void Remove(string commandName)
         {
             string commandFileName = commandName +
-                (PlatformHelper.IsWindows ? ".cmd" : string.Empty);
+                (_isWindows ? ".cmd" : string.Empty);
```
Comment (moozzyk) [inline src/Microsoft.Framework.PackageManager/Install/AppCommandsFolderRepository.cs]: nit: use var (same below)

File: src/Microsoft.Framework.PackageManager/Install/AppCommandsFolderRepository.cs [hunk 12]
```diff
@@ -76,7 +78,7 @@ public void Load()
         {
             _commands = new Dictionary<string, NuGet.PackageInfo>();
 
-            string pathFilter = PlatformHelper.IsWindows ? "*.cmd" : "*.*";
+            string pathFilter = _isWindows ? "*.cmd" : "*.*";
```
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Install/AppCommandsFolderRepository.cs]: Let's do some cleanup by the way. Use var here.

File: src/Microsoft.Framework.PackageManager/Program.cs [hunk 13]
```diff
@@ -412,7 +412,7 @@ public int Main(string[] args)
                     }
 
                     var command = new DependencyListCommand(options, _environment.RuntimeFramework);
-                    return command.Execute();
+                    return command.Execute(_hostServices);
```
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Program.cs]: I suggest we follow one consistent pattern in this file. That is, for class `XYZCommand`, we pass environment info/services it needs to its ctor and `XYZCommand.Execute()` should take no parameter.

File: src/Microsoft.Framework.PackageManager/Program.cs [hunk 14]
```diff
@@ -518,7 +519,7 @@ public int Main(string[] args)
                     command.InPlace = optInPlace.HasValue();
                     command.Framework = optFramework.Value();
 
-                    var success = command.ExecuteCommand();
+                    var success = command.ExecuteCommand(RuntimeEnvironmentHelper.IsMono(_runtimeEnv));
```
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Program.cs]: Same here. Runtime info should be passed to ctor and ExecuteCommand() should take no parameter.

Please help do some cleanup BTW, let's rename this `WrapCommand.ExecuteCommand()` to `WrapCommand.Execute()` to remove redundancy in the name.

File: src/Microsoft.Framework.PackageManager/Utils/VersionUtils.cs [hunk 15]
```diff
@@ -25,7 +26,7 @@ private static string GetActiveRuntimeName()
 
             if (!string.IsNullOrEmpty(pathVariable))
             {
-                string dnuExecutable = PlatformHelper.IsWindows ? "dnu.cmd" : "dnu";
+                string dnuExecutable = RuntimeEnvironmentHelper.IsWindows(CallContextServiceLocator.Locator.ServiceProvider) ? "dnu.cmd" : "dnu";
```
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Utils/VersionUtils.cs]: If we can do this, why do we want to pass `IServiceProvider/IRuntimeEnvironment` around? If we have a global instance of service provider, can we just have a property `RuntimeEnvironmentHelper.IsWindows`, in which the boolean is lazily evaluated with this global service provider?

**Comment by BrennanConroy** [comment]
After talking with David, we would like to do something similar to what Platformhelper did, so my proposed solution is:
```
internal static class RuntimeEnvironmentHelper
{
    private static Lazy<bool> _isMono = new Lazy<bool>(() => ((IRuntimeEnvironment)Services.GetService(typeof(IRuntimeEnvironment))).RuntimeType == "Mono");
    private static Lazy<bool> _isWindows = new Lazy<bool>(() => ((IRuntimeEnvironment)Services.GetService(typeof(IRuntimeEnvironment))).RuntimeType == "Mono");

    public static IServiceProvider Services { private get; set; }

    public static bool IsWindows
    {
        get
        {
            if(Services == null)
            {
                throw new NullReferenceException("Services hasn't been assigned yet");
            }

            return _isWindows.Value;
        }
    }

    public static bool IsMono
    {
        get
        {
            if (Services == null)
            {
                throw new NullReferenceException("Services hasn't been assigned yet");
            }

            return _isMono.Value;
        }
    }
}
```

We will need to initialize the Services at the beginning, and then we can use the Helper just like we did PlatformHelper before, which will keep the code a lot cleaner.

**Comment by ChengTian** [comment]
:+1:

**Comment by ChengTian** [comment]
How do you initialize Services? Make `RuntimeEnvironmentHelper.Services` lazily initialized from `CallContextServiceLocator.Locator.ServiceProvider`?

**Comment by BrennanConroy** [comment]
Oh ya, that's a good idea

File: src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+using System;
+
+namespace Microsoft.Framework.Runtime
+{
+    internal static class RuntimeEnvironmentHelper
+    {
+        private static Lazy<bool> _isMono = new Lazy<bool>(() => ((IRuntimeEnvironment)Services.Value.GetService(typeof(IRuntimeEnvironment))).RuntimeType == "Mono");
+        private static Lazy<bool> _isWindows = new Lazy<bool>(() => ((IRuntimeEnvironment)Services.Value.GetService(typeof(IRuntimeEnvironment))).OperatingSystem == "Windows");
+
+        public static Lazy<IServiceProvider> Services = new Lazy<IServiceProvider>(() => Infrastructure.CallContextServiceLocator.Locator.ServiceProvider);
```
Comment (BrennanConroy) [inline src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs]: Change to private

File: src/Microsoft.Framework.PackageManager/Program.cs [hunk 17]
```diff
@@ -529,7 +527,7 @@ public int Main(string[] args)
 
         private Reports CreateReports(bool verbose, bool quiet)
         {
-            var useConsoleColor = _runtimeEnv.OperatingSystem == "Windows";
+            var useConsoleColor = false;
```
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Program.cs]: Why? Shouldn't it be
```c#
var useConsoleColor = RuntimeEnvironmentHelper.IsWindows
```

Comment (BrennanConroy) [inline src/Microsoft.Framework.PackageManager/Program.cs]: Oh ya, well the AnsiConsole will be changed in another PR before this goes out, so I didn't change this.

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynProjectReference.cs [hunk 18]
```diff
@@ -270,7 +270,7 @@ private static bool SupportsPdbGeneration()
         {
             try
             {
-                if (PlatformHelper.IsMono)
```
Comment (BrennanConroy) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynProjectReference.cs]: Should this be !IsWindows?

Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynProjectReference.cs]: Sure, unless the CLR team has a solution to generate PDBs on non windows

Comment (moozzyk) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynProjectReference.cs]: nit: I don't know if it makes a lot of difference but why this check is inside the try/catch block?

File: src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs [hunk 19]
```diff
@@ -30,7 +30,7 @@ public class WrapCommand
         public string Framework { get; set; }
         public Reports Reports { get; set; }
 
-        public bool ExecuteCommand()
+        public bool Execute()
```
Comment (troydai) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: The change of the method name seems not to be related to this PR.

File: src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+        private static Lazy<bool> _isWindows = new Lazy<bool>(() => ((IRuntimeEnvironment)Services.Value.GetService(typeof(IRuntimeEnvironment))).OperatingSystem == "Windows");
+
+        private static Lazy<IServiceProvider> Services = new Lazy<IServiceProvider>(() => Infrastructure.CallContextServiceLocator.Locator.ServiceProvider);
+
+        public static bool IsWindows
+        {
+            get
+            {
+                return _isWindows.Value;
+            }
```
Comment (troydai) [inline src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs]: Put these in one line to save space.

File: src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs [hunk 21]
```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Framework.Runtime
+{
+    internal static class RuntimeEnvironmentHelper
+    {
+        private static Lazy<bool> _isMono = new Lazy<bool>(() => ((IRuntimeEnvironment)Services.Value.GetService(typeof(IRuntimeEnvironment))).RuntimeType == "Mono");
```
Comment (troydai) [inline src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs]: This line is too long. Break it.

File: src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+        private static Lazy<bool> _isWindows = new Lazy<bool>(() => ((IRuntimeEnvironment)Services.Value.GetService(typeof(IRuntimeEnvironment))).OperatingSystem == "Windows");
```
Comment (troydai) [inline src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs]: Break the line.

File: src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+        private static Lazy<IServiceProvider> Services = new Lazy<IServiceProvider>(() => Infrastructure.CallContextServiceLocator.Locator.ServiceProvider);
```
Comment (troydai) [inline src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs]: Break the line.

**Comment by troydai** [comment]
:watch:

File: src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs (hunk 19 above)
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: We can keep this. In previous approach, @BrennanConroy 's changes conflict with mine so I asked him to make this change. All other `XYZCommand.ExecuteCommand()` has been renamed to `XYZCommand.Execute()` by me. This is the only one left.

File: test/Microsoft.Framework.PackageManager.FunctionalTests/Microsoft.Framework.PackageManager.FunctionalTests.xproj [hunk 24]
```diff
@@ -13,8 +13,5 @@
   <PropertyGroup>
     <SchemaVersion>2.0</SchemaVersion>
   </PropertyGroup>
-  <ItemGroup>
-    <Service Include="{82a7f48d-3b50-4b1e-b82e-3ada8210c358}" />
-  </ItemGroup>
```
Comment (dougbu) [inline test/Microsoft.Framework.PackageManager.FunctionalTests/Microsoft.Framework.PackageManager.FunctionalTests.xproj]: Don't remove this. It's a test service reference and helps VS load the solution faster.

File: src/Microsoft.Framework.PackageManager/Utils/FileOperationUtils.cs [hunk 25]
```diff
@@ -38,12 +37,6 @@ private static void DeleteRecursive(string deletePath)
 
         public static bool MarkExecutable(string file)
         {
-            if (PlatformHelper.IsWindows)
```
Comment (dougbu) [inline src/Microsoft.Framework.PackageManager/Utils/FileOperationUtils.cs]: Why attempt to execute `chmod` on Windows?

File: src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs [hunk 26]
```diff
@@ -47,7 +47,7 @@ public bool ExecuteCommand()
             var extension = Path.GetExtension(InputFilePath);
             if (string.Equals(extension, ".csproj", StringComparison.OrdinalIgnoreCase))
             {
-                if (PlatformHelper.IsMono)
+                if (!RuntimeEnvironmentHelper.IsWindows)
```
Comment (dougbu) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: Curious: What are there rules by which you've chosen `IsMono` -> `IsMono` in some places and `IsMono` -> `!IsWindows` in others?

Comment (BrennanConroy) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: IsMono was used incorrectly in most places because we were only doing stuff on Windows and on Mono, now that we are getting cross-plat CoreCLR support in we need to fix all of these incorrect IsMono's

Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: Yeah, previously without the utils added by this PR, we have no way to tell the OS we are running on. So we assumed `IsMono == non-Windows`. Now we are fixing this.

File: src/Microsoft.Framework.PackageManager/Utils/FileOperationUtils.cs (hunk 25 above)
Comment (BrennanConroy) [inline src/Microsoft.Framework.PackageManager/Utils/FileOperationUtils.cs]: I removed this a while ago during an earlier iteration because I noticed everytime we call this function we already did the check for Windows. Would it be better to keep the check in here and remove it from the places we call it?

Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Utils/FileOperationUtils.cs]: Sounds good. This is a shared util method and putting OS check in it makes code DRYer.

Comment (dougbu) [inline src/Microsoft.Framework.PackageManager/Utils/FileOperationUtils.cs]: Definitely safer and results in less code duplication. I'd also have to change #1931 if this went in first :smiling_imp:

File: src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs (hunk 26 above)
Comment (dougbu) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: thx

File: src/Microsoft.Framework.PackageManager/Utils/FileOperationUtils.cs (hunk 25 above)
Comment (moozzyk) [inline src/Microsoft.Framework.PackageManager/Utils/FileOperationUtils.cs]: Add an assert.

**Comment by ChengTian** [comment]
Please address the last few comments and :shipit:

**Comment by BrennanConroy** [comment]
We can't merge this until we use a version of mono that doesn't null-ref

File: src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs (hunk 23 above)
Comment (ChengTian) [inline src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs]: `_services`?

File: src/dnx.host/Bootstrapper.cs [hunk 27]
```diff
@@ -76,6 +76,7 @@ public Task<int> RunAsync(List<string> args, IRuntimeEnvironment env)
                 serviceProvider.Add(typeof(IRuntimeEnvironment), env);
 
                 CallContextServiceLocator.Locator.ServiceProvider = serviceProvider;
+                var mono = RuntimeEnvironmentHelper.IsMono;
```
Comment (ChengTian) [inline src/dnx.host/Bootstrapper.cs]: Why are we adding this? Seems we don't use this newly added variable

Comment (BrennanConroy) [inline src/dnx.host/Bootstrapper.cs]: I should add a comment. This is my proposed temporary fix for mono null-ref. If we grab the value here it wont fail later.

Comment (davidfowl) [inline src/dnx.host/Bootstrapper.cs]: Add the comment and add an initialize method that sets it up once instead of all the lazy crazyness

File: src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs (hunk 26 above)
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: I'm not sure this one is corect

File: src/Microsoft.Framework.Runtime/FrameworkReferenceResolver.cs [hunk 28]
```diff
@@ -185,7 +185,8 @@ private static FrameworkInformation GetFrameworkInformation(FrameworkName target
 
             // Skip this on mono since it has a slightly different set of reference assemblies at a different
             // location
-            if (!PlatformHelper.IsMono && FrameworkDefinitions.TryPopulateFrameworkFastPath(targetFramework.Identifier, targetFramework.Version, referenceAssembliesPath, out frameworkInfo))
+            var isMono = RuntimeEnvironmentHelper.IsMono;
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/FrameworkReferenceResolver.cs]: Why did you declare a variable?

File: src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs (hunk 26 above)
Comment (BrennanConroy) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: Ok, who is the owner of Wrap?

Comment (BrennanConroy) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: Wei said this is correct because wrap calls MSBuild

Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: MSBuild exists on windows and since it shells out it will work with CoreCLR on windows.

Comment (BrennanConroy) [inline src/Microsoft.Framework.PackageManager/Commands/WrapCommand.cs]: Ok? How is this wrong though

**Comment by BrennanConroy** [comment]
Looks like Mono has fixed the Execution Context issue in the released bits, will run some more to be sure.

File: src/Microsoft.Framework.PackageManager/Publish/PublishRuntime.cs [hunk 29]
```diff
@@ -43,7 +43,7 @@ public bool Emit(PublishRoot root)
 
             new PublishOperations().Copy(_runtimePath, TargetPath);
 
-            if (PlatformHelper.IsMono)
+            if (!RuntimeEnvironmentHelper.IsWindows)
```
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Publish/PublishRuntime.cs]: Do we want something like `RuntimeEnvironmentHelper.IsUnixLike` (which is `IsLinux || IsDarwin`)?

File: test/Microsoft.Framework.CommonTestUtils/TestUtils.cs [hunk 30]
```diff
@@ -403,7 +403,7 @@ private static bool IsWindows()
             var p = (int)Environment.OSVersion.Platform;
             return (p != 4) && (p != 6) && (p != 128);
 #else
-            return PlatformHelper.IsWindows;
+            return RuntimeEnvironmentHelper.IsWindows;
```
Comment (davidfowl) [inline test/Microsoft.Framework.CommonTestUtils/TestUtils.cs]: Use the runtime helper for everything, not just for DNXCORE50

File: src/Microsoft.Framework.Runtime/DependencyManagement/GacDependencyResolver.cs [hunk 31]
```diff
@@ -17,7 +17,7 @@ public class GacDependencyResolver : IDependencyProvider, ILibraryExportProvider
 
         public IEnumerable<string> GetAttemptedPaths(FrameworkName targetFramework)
         {
-            if (PlatformHelper.IsMono)
+            if (!RuntimeEnvironmentHelper.IsWindows)
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/DependencyManagement/GacDependencyResolver.cs]: Add the mono check as well (windows on mono should fail)

File: src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs [hunk 32]
```diff
@@ -0,0 +1,29 @@
...
+{
+    internal static class RuntimeEnvironmentHelper
+    {
+        private static Lazy<bool> _isMono = new Lazy<bool>(() =>
+            ((IRuntimeEnvironment)_services.Value.GetService(typeof(IRuntimeEnvironment))).RuntimeType == "Mono");
+
+        private static Lazy<bool> _isWindows = new Lazy<bool>(() =>
+            ((IRuntimeEnvironment)_services.Value.GetService(typeof(IRuntimeEnvironment))).OperatingSystem == "Windows");
+
+        private static Lazy<IServiceProvider> _services = new Lazy<IServiceProvider>(() =>
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Sources/Impl/RuntimeEnvironmentHelper.cs]: Make this `Lazy<IRuntimeEnvironment>`

File: src/Microsoft.Framework.PackageManager/Publish/PublishRuntime.cs (hunk 29 above)
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Publish/PublishRuntime.cs]: That's interesting. We should think about that.

**Comment by davidfowl** [comment]
:shipit: After those few comments. Make sure that older versions of mono don't start null reffing all over the place because of a broken call context. This change needs to work without that fix.

**Comment by muratg** [comment]
When can we get this in?  @ChengTian's fix is waiting on this one.

**Comment by BrennanConroy** [comment]
@davidfowl 
Any more comments? Or ship?

**Comment by davidfowl** [comment]
@BrennanConroy I assume you know why appveyor is failing and that this change didn't actually regress anything.

If that's the case then :shipit:

**Comment by BrennanConroy** [comment]
Ya, appveyor failed because of the System.Reflection.Primitives not being found.

**Comment by kichalla** [comment]
Several other repos have their own copy of `PlatformHelper`. So is this the new recommended way of checking for platform? @BrennanConroy 

https://github.com/search?q=user%3Aaspnet+platformhelper&ref=searchresults&type=Code&utf8=%E2%9C%93

cc: @bricelam

**Comment by victorhurdugaci** [comment]
@kichalla no. you should use `IApplicationEnvironment`
